feat(spanner): support of client-level custom retry settings

hengfengli opened the pull request:
Changes included are: 

- Add a field `CallOptions` in `ClientConfig`.
- Add a field `callOptions` in `sessionClient`.
- Add a method `mergeCallOptions` to merge two CallOptions.
- Update `nextClient` in sessionclient.go so that if a custom CallOptions is provided, merge it with the default one. 

Fixes #2594

olavloite commented on spanner/sessionclient.go:
```diff
@@ -272,5 +276,31 @@ func (sc *sessionClient) nextClient() (*vkit.Client, error) {
+
+// mergeCallOptions merges two CallOptions into one and the first argument has
+// a lower order of precedence than the second one.
+func mergeCallOptions(low *vkit.CallOptions, high *vkit.CallOptions) *vkit.CallOptions {
```
nit: I find the `low` and `high` variable names here a little confusing. It kind of indicates that it will give precedence to the highest or lowest value, while it is just `a` and `b`, where `b` takes precedence. So could we rename it to either `a` and `b` or `first` and `second` (or something like that)?

hengfengli replied:
That's a good catch. Originally, I used `src` and `dest`, but then I feel it is not quite right and changed to `low` and `high`. I will go for `a` and `b`.

olavloite commented on spanner/sessionclient_test.go:
```diff
@@ -373,3 +375,77 @@ func TestClientIDGenerator(t *testing.T) {
+		t.Fatalf("merged CallOptions is incorrect: got %v, want %v", got, want)
+	}
+
+	merged.CreateSession[1].Resolve(cs)
```
I'm not very familiar with the exact working of these options in Go. When exactly would the `merged.CreateSession[1]` setting be used instead of the `merged.CreateSession[0]` setting?

hengfengli replied:
https://github.com/googleapis/google-cloud-go/blob/06b6069396a620973802c129f2b58056a079cec5/spanner/apiv1/spanner_client.go#L308-L311

In the generated method, you can see that `opts = [c.CallOptions.CreateSession, opts...]`. So the passed-in opts have higher order of precedence. 

Therefore, `merged.CreateSession[1]` will override `merged.CreateSession[0]`. 

My guess is that it will iterate all `gax.CallOption` in `merged.CreateSession` and call `Resolve` with `gax.CallSettings`. Only the effect of the last one remains. Then, it uses `gax.CallSettings` for calling gRPC.

olavloite commented on spanner/client_test.go:
```diff
@@ -1757,6 +1759,43 @@ func TestClient_WithGRPCConnectionPoolAndNumChannels_Misconfigured(t *testing.T)
+	}
+
+	// This is the custom retry setting.
+	c.CallOptions.CreateSession[1].Resolve(cs)
```
Is this the setting that will actually be used in this case? Is it in that case necessary to keep the other settings still in `c.CallOptions.CreateSession[0]`?

hengfengli replied:
As I described in the previous comment, `c.CallOptions.CreateSession[1]` has higher order of precedence. Yes, it is not necessary to keep `c.CallOptions.CreateSession[0]`. But I have two reasons to do it in this way: 

- Follow the same pattern in `spanner_client.go`: `opts = append(c.CallOptions.CreateSession[0:len(c.CallOptions.CreateSession):len(c.CallOptions.CreateSession)], opts...)`
- Simplify the merging logic: we don't need to say `if b is not empty, use b`, `else if a is not empty, use a`, and `if both empty, leave it empty`. We can just merge `a` and `b` in the right order.

olavloite replied:
Sounds reasonable to me. I do think we should consider adding some kind of end-to-end test for this, as we are relying on what is effectively an undocumented feature of the generated client that might silently change in the future. That end-to-end test could probably be created using the mock server, and does not need to be a full integration test.

hengfengli commented:
The default CallOptions are set in: 

https://github.com/googleapis/google-cloud-go/blob/06b6069396a620973802c129f2b58056a079cec5/spanner/apiv1/spanner_client.go#L256-L261

We override the default CallOptions in: 

https://github.com/googleapis/google-cloud-go/blob/c455fe2c4131ec763ca6181ede26ecd924ead693/spanner/sessionclient.go#L279-L281

`client.CallOptions` is the default CallOptions and `sc.callOptions` is the custom CallOptions. 

Assume `client.CallOptions` is: 

```bash
client.CallOptions = {
   CreateSession: [opt1],
   BatchCreateSessions: [op2],
   GetSession: [],
}
``` 

and `sc.callOptions` is:

```bash
sc.callOptions = {
   CreateSession: [opt3],
   BatchCreateSessions: [],
   GetSession: [op5],
}
``` 

Then, the merged one will become: 

```bash
sc.callOptions = {
   CreateSession: [opt1, opt3],
   BatchCreateSessions: [opt2],
   GetSession: [op5],
}
```

The final effect would be `CreateSession: opt3`, `BatchCreateSessions: opt2`, `GetSession: opt5`. The last option in each array takes effect.

hengfengli commented:
I dig a bit more into the code and found: 

https://github.com/googleapis/gax-go/blob/be11bb253a768098254dc71e95d1a81ced778de3/v2/invoke.go#L41-L49

It actually does what I think. It iterates all CallOptions and calling `Resolve()` with `CallSettings`. 

`Resolve()` just overrides the `Retry` field of `CallSettings`:

https://github.com/googleapis/gax-go/blob/be11bb253a768098254dc71e95d1a81ced778de3/v2/call_option.go#L57-L59

Therefore, only the last one will be remained in the `Retry` field. 

But I am happy to add an end-to-end test for this.

hengfengli commented:
@olavloite , I have added an end-to-end test. Please take a look. 

I learned an interesting fact that the gax retrying does not actually work for `ExecuteStreamingSql()`. When we call `ExecuteStreamingSql()`, we actually get a receiver, which does not receive any result yet. It only gets results when we call `iter.Next()` where the gax retrying will not work. 

Basically, what I learned is: 

* Create an iterator: create a spannerExecuteStreamingSqlClient (gax retrying) - sending messages to the stream. 
* Iterate an iterator: make a ExecuteStreamingSql request (no gax retrying) - receiving results from the stream. 

Originally, I tried to set a custom retry setting for ExecuteStreamingSql(), but it doesn't work. So I changed it to use ExecuteSql().

olavloite commented:
> @olavloite , I have added an end-to-end test. Please take a look.
> 
> I learned an interesting fact that the gax retrying does not actually work for `ExecuteStreamingSql()`. When we call `ExecuteStreamingSql()`, we actually get a receiver, which does not receive any result yet. It only gets results when we call `iter.Next()` where the gax retrying will not work.
> 
> Basically, what I learned is:
> 
> * Create an iterator: create a spannerExecuteStreamingSqlClient (gax retrying) - sending messages to the stream.
> * Iterate an iterator: make a ExecuteStreamingSql request (no gax retrying) - receiving results from the stream.
> 
> Originally, I tried to set a custom retry setting for ExecuteStreamingSql(), but it doesn't work. So I changed it to use ExecuteSql().

@hengfengli Thanks for adding the end to end test. This looks good to me.

The retry logic for `ExecuteStreamingSql` is currently [hand-written](https://github.com/googleapis/google-cloud-go/blob/a31e954c9776cea53bcf218b2f94538d74e1b71c/spanner/read.go#L406), partly also because it should use the last seen resume token that was received from the server. So if we would want to allow users to specify custom retry settings for queries as well, then we would have to [make this retryer configurable](https://github.com/googleapis/google-cloud-go/blob/a31e954c9776cea53bcf218b2f94538d74e1b71c/spanner/read.go#L460).
